### PR TITLE
[dev-server] support building hermes bytecode bundle for ios

### PR DESCRIPTION
**Pull request by Kudo**
# Why

support building hermes bytecode bundle for ios

# How

- add ios and shared `jsEngine` support to `isEnableHermesManaged` 
- some refactoring to make `maybeAddHermesBundleAsync` clear
  = move sdk check into HermesBundler.ts
  = move throwing error of `maybeInconsistentEngine*` into HermesBundler.ts
- add `maybeInconsistentEngineIosAsync` to support Podfile.properties.json

# Test Plan

### new unit tests

```
 PASS   @expo/dev-server  src/__tests__/HermesBundler-test.ts
  isEnableHermesManaged
    ✓ should support shared jsEngine key
    ✓ platform jsEngine should override shared jsEngine
    ✓ should exclude old SDK (18 ms)
  maybeThrowFromInconsistentEngineAsync - ios
    ✓ should resolve if ":hermes_enabled => true" in Podfile and "jsEngine: 'hermes'" in app.json (1 ms)
    ✓ should throw if ":hermes_enabled => true" in Podfile but no "jsEngine: 'hermes'" in app.json
    ✓ should throw if (":hermes_enabled => false" or not existed in Podfile) and "jsEngine: 'hermes'" in app.json (1 ms)
    ✓ should resolve if "expo.jsEngine": 'hermes' in Podfile.properties.json and "jsEngine: 'hermes'" in app.json
    ✓ should handle the inconsistency between Podfile and Podfile.properties.json
```

### updated unit tests (mostly for `maybeThrowFromInconsistentEngineAsync` interface change)

```
  maybeThrowFromInconsistentEngineAsync - common
    ✓ should resolve for managed project
  maybeThrowFromInconsistentEngineAsync - android
    ✓ should resolve if "enableHermes: true" in app/build.gradle and "jsEngine: 'hermes'" in app.json (3 ms)
    ✓ should throw if "enableHermes: false" in app/build.gradle and "jsEngine: 'hermes'" in app.json (1 ms)
    ✓ should resolve if "expo.jsEngine=hermes" in gradle.properties and "jsEngine: 'hermes'" in app.json (1 ms)
    ✓ should resolve if "expo.jsEngine=hermes" in gradle.properties but no "jsEngine: 'hermes'" in app.json
    ✓ should resolve for old sdk bare project
    ✓ should handle the inconsistency between app/build.gradle and gradle.properties
```

### manually test

```sh
$ expo init extest # select managed
# add `jsEngine: "hermes"` and `sdkVersion: "UNVERSIONED"` to app.json
$ EXPO_SKIP_MANIFEST_VALIDATION_TOKEN=1 /path/to/expo-cli/packages/expo-cli/bin/expo.js export -p https://example.com/
```

### CODE REVIEW

File: packages/dev-server/src/MetroDevServer.ts [hunk 1]
```diff
@@ -198,40 +183,32 @@ export async function bundleAsync(
       );
       bundleOutput.hermesBytecodeBundle = hermesBundleOutput.hbc;
       bundleOutput.hermesSourcemap = hermesBundleOutput.sourcemap;
+
+      if (platform === 'ios') {
+        options.logger.warn(
+          { tag: 'expo' },
+          `❗️ ${platformTag} Over-the-Air updates with Hermes engine may violate App Store Review Guidelines. Publish the bundle at risk.`
```
Comment (EvanBacon) [inline packages/dev-server/src/MetroDevServer.ts:190]: @ide @brentvatne

File: packages/dev-server/src/HermesBundler.ts [hunk 2]
```diff
@@ -131,6 +168,41 @@ async function maybeInconsistentEngineAndroidAsync(
...
+    if (!isPropsReference && isHermesManaged !== isHermesBare) {
+      return true;
+    }
+  }
+
+  // Check Podfile.properties.json from prebuild template
+  const podfilePropertiesPath = path.join(projectRoot, 'ios', 'Podfile.properties.json');
+  if (fs.existsSync(podfilePropertiesPath)) {
+    try {
+      const props = JSON.parse(await fs.readFile(podfilePropertiesPath, 'utf8'));
```
[the remark this answers is not in the document]
Comment (Kudo) [reply packages/dev-server/src/HermesBundler.ts]: is this likely to be what you mean or even placing `fs.existsSync` check into the function? https://github.com/expo/expo-cli/pull/3814/commits/261f68e45757cb794c6f7f7026b286753cb03a23

[the remark this answers is not in the document]
Comment (EvanBacon) [reply packages/dev-server/src/HermesBundler.ts]: Ya that looks good

[the remark this answers is not in the document]
Comment (Kudo) [reply packages/dev-server/src/HermesBundler.ts]: thanks! please let me know if there are any other comments 😄

**Review by EvanBacon — APPROVED** [review]
lgtm

**Pull request merged**